vsphere/resource_vsphere_host.go

```diff
@@ -0,0 +1,683 @@
+	}
+
+	hamRef := hostProps.ConfigManager.HostAccessManager.Reference()
+	ham := NewHostAccessManager(client.Client, hamRef)
+	err = ham.ChangeLockdownMode(context.TODO(), lockdownMode)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func modifyMaintenanceMode(d *schema.ResourceData, meta, old, new interface{}) error {
+	client := meta.(*VSphereClient).vimClient
+	hostID := d.Id()
+
+	host, err := hostsystem.FromID(client, hostID)
+	if err != nil {
+		return err
+	}
+
+	maintenanceMode := new.(bool)
+	var maintErr error
+	if maintenanceMode {
+		maintErr = hostsystem.EnterMaintenanceMode(host, int(defaultAPITimeout), true)
+	} else {
+		maintErr = hostsystem.ExitMaintenanceMode(host, int(defaultAPITimeout))
+	}
+	if maintErr != nil {
+		return maintErr
+	}
+	return nil
+}
+
+func modifyLicense(d *schema.ResourceData, meta, old, new interface{}) error {
+	client := meta.(*VSphereClient).vimClient
+	lm := license.NewManager(client.Client)
+	lam, err := lm.AssignmentManager(context.TODO())
+	if err != nil {
+		return err
+	}
+	lam.Update(context.TODO(), d.Id(), new.(string), "")
+	return nil
+}
+
+func modifyCluster(d *schema.ResourceData, meta, old, new interface{}) error {
+	client := meta.(*VSphereClient).vimClient
+	hostID := d.Id()
+	newClusterID := new.(string)
+
+	newCluster, err := clustercomputeresource.FromID(client, newClusterID)
+	if err != nil {
+		log.Printf("[DEBUG] Error while searching new cluster %s. Error: %s.", newClusterID, err)
+		return err
+	}
+
+	hs, err := hostsystem.FromID(client, hostID)
+	if err != nil {
+		return err
+	}
+
+	err = hostsystem.EnterMaintenanceMode(hs, int(defaultAPITimeout), true)
+	if err != nil {
+		return fmt.Errorf("error while putting host to maintenance mode: %s", err.Error())
+	}
+
+	task, err := newCluster.MoveInto(context.TODO(), hs)
+	if err != nil {
+		return err
+	}
+	task.Wait(context.TODO())
+
+	err = hostsystem.ExitMaintenanceMode(hs, int(defaultAPITimeout))
+	if err != nil {
+		return fmt.Errorf("error while taking host out of maintenance mode: %s", err.Error())
+	}
+
+	var to mo.Task
+	err = task.Properties(context.TODO(), task.Reference(), nil, &to)
+	if err != nil {
+		log.Printf("[DEBUG] Failed while getting task results: %s", err)
+		return err
+	}
+
+	if to.Info.State != "success" {
+		return fmt.Errorf("Error while moving host to new cluster (%s): %s", newClusterID, to.Info.Error)
+	}
+
+	return nil
+}
+
+func handleReconnect(d *schema.ResourceData, meta interface{}) error {
+	hostID := d.Id()
+	client := meta.(*VSphereClient).vimClient
+	host := object.NewHostSystem(client.Client, types.ManagedObjectReference{Type: "HostSystem", Value: d.Id()})
+	hcs := types.HostConnectSpec{
+		HostName:      d.Get("hostname").(string),
+		UserName:      d.Get("username").(string),
+		Password:      d.Get("password").(string),
+		SslThumbprint: d.Get("thumbprint").(string),
+		Force:         d.Get("force").(bool),
+	}
+
+	task, err := host.Reconnect(context.TODO(), &hcs, nil)
```

Comment:
Same as above for adding a function since you're dealing with a context and waiting on tasks.